src/cdk-experimental/ui-patterns/behaviors/event-manager/event-manager.ts

```diff
@@ -0,0 +1,192 @@
+  handler: (event: T) => Promise<boolean | void>;
+}
+
+/**
+ * Bit flag representation of the possible modifier keys that can be present on an event.
+ */
+export enum ModifierKey {
+  None = 0,
+  Ctrl = 0b1,
+  Shift = 0b10,
+  Alt = 0b100,
+  Meta = 0b1000,
+}
+
+/**
+ * Abstract base class for all event managers.
+ *
+ * Event managers are designed to normalize how event handlers are authored and create a safety net
+ * for common event handling gotchas like remembering to call preventDefault or stopPropagation.
+ */
+export abstract class EventManager<T extends Event> {
+  private _submanagers: EventManager<T>[] = [];
+
+  protected configs: EventHandlerConfig<T>[] = [];
+  protected beforeFns: ((event: T) => void)[] = [];
+  protected afterFns: ((event: T) => void)[] = [];
+
+  protected defaultHandlerOptions: EventHandlerOptions = {
+    preventDefault: false,
+    stopPropagation: false,
+  };
+
+  constructor(defaultHandlerOptions?: Partial<EventHandlerOptions>) {
+    this.defaultHandlerOptions = {
+      ...this.defaultHandlerOptions,
+      ...defaultHandlerOptions,
+    };
+  }
+
+  /**
+   * Composes together multiple event managers into a single event manager that delegates to the
+   * individual managers.
+   */
+  static compose<T extends Event>(...managers: EventManager<T>[]) {
+    const composedManager = new GenericEventManager<T>();
+    composedManager._submanagers = managers;
+    return composedManager;
+  }
+
+  /**
+   * Runs any handlers that have been configured to handle this event. If multiple handlers are
+   * configured for this event, they are run in the order they were configured. Returns
+   * `true` if the event has been handled, otherwise returns `undefined`.
+   *
+   * Note: the use of `undefined` instead of `false` in the unhandled case is necessary to avoid
+   * accidentally preventing the default behavior on an unhandled event.
+   */
+  async handle(event: T): Promise<true | undefined> {
+    if (!this.isHandled(event)) {
+      return undefined;
+    }
+    for (const fn of this.beforeFns) {
+      fn(event);
+    }
+    for (const submanager of this._submanagers) {
+      await submanager.handle(event);
+    }
+    for (const config of this.getConfigs(event)) {
+      await config.handler(event);
+      if (config.stopPropagation) {
+        event.stopPropagation();
+      }
+      if (config.preventDefault) {
+        event.preventDefault();
+      }
+    }
+    for (const fn of this.afterFns) {
+      fn(event);
+    }
+    return true;
+  }
+
+  /**
+   * Configures the event manager to run a function immediately before it as about to handle
+   * any event.
+   */
+  beforeHandling(fn: (event: T) => void): this {
+    this.beforeFns.push(fn);
+    return this;
+  }
+
+  /**
+   * Configures the event manager to run a function immediately after it handles any event.
+   */
+  afterHandling(fn: (event: T) => void): this {
```

Comment:
What's the use case for using `beforeHandling` over `afterHandling`?

Reply:
The idea was you'd set up a manager to handle, e.g. all the navigation events, then you could pass that manager to the selection behavior, and the selection behavior could add logic that runs before/after any navigation event is handled. At least that's what we had in some version of the prototype - @wagnermaciel did anything like that make it into this final version? if not maybe we can just kill these methods

Reply:
Nope. I'll remove the methods